### PR TITLE
feat(sozo): add ability to migrate the world

### PR DESCRIPTION
**Pull request by lambda-0x**
fix: #1671
fix: DOJ-268

- [x] make sure its works as expected
- [x] add `original_base_class_hash` for `DojoContract`s

### CODE REVIEW

**Comment by glihm** [comment]
As a note, the saya test is fixed on main. :+1:

**Comment by glihm** [comment]
@lambda-0x I've made a test where:
1. migrate the spawn and move project without modification.
2. change a bit the world to have a new class hash.
3. migrate again, and the world is not updated in-place.
4. the base class hash wasn't modify, we should expect a world upgrade, right?

**Comment by lambda-0x** [comment]
@glihm would need to create an `overlays/world.toml` file with the `original_class_hash` set to the original class hash of the world

**Comment by lambda-0x** [comment]
(will fix the conflicts once changes are reviewed)

**Comment by glihm** [comment]
> @glihm would need to create an `overlays/world.toml` file with the `original_class_hash` set to the original class hash of the world

an important point to keep for documentation, the `name` field is required.

**Comment by glihm** [comment]
@lambda-0x it seems like if the world get back to the original class hash after being upgraded, it does not upgrade to ensure the class hash is restored.

1. Deploy the world
2. Modify the world + migrate + overlay -> it upgrades correctly
3. Restore the world as it was in 1, and migrate + remove overlay -> it does not upgrade

We need a big refactoring on the code to organize the testing on all of that. 👍

File: crates/sozo/ops/src/migration/mod.rs [hunk 1]
```diff
@@ -411,68 +420,54 @@ where
...
-                } else {
-                    (None, None)
-                };
-
-            ui.print_sub(format!("Contract address: {:#x}", world.contract_address));
-
-            let offline = ws.config().offline();
+            if world.diff.local_class_hash != world.diff.original_class_hash
+                && world.diff.original_class_hash
+                    == world.diff.remote_class_hash.unwrap_or_default()
```
Comment (glihm) [inline crates/sozo/ops/src/migration/mod.rs:425]: Not sure about this condition and it's related to my previous comment.

To here, we want to actually upgrade anytime local class hash is different from the remote one, independently of the original class hash match.

File: crates/dojo-world/src/migration/mod.rs [hunk 2]
```diff
@@ -35,6 +33,16 @@ pub struct DeployOutput {
     pub block_number: Option<u64>,
     pub contract_address: FieldElement,
     pub declare: Option<DeclareOutput>,
+    // base class hash at time of deployment
+    pub base_class_hash: FieldElement,
+}
+
+#[derive(Clone, Debug)]
+pub struct UpgradeOutput {
```
Comment (glihm) [inline crates/dojo-world/src/migration/mod.rs:41]: We should see how to use that on the dojo contracts. Currently even if upgraded, we return a `DeployOutput` struct only.

**Comment by glihm** [comment]
@lambda-0x would appreciate a double check if you have the bandwidth. 👍

**Comment by glihm** [comment]
@lambda-0x I'll merge for now, we should catch up for tests architecture once you're available. 👍

Related to #1684.

**Review by glihm — APPROVED** [review]

**Pull request merged**